packages/local-relay/src/RelayServer.ts

```diff
@@ -0,0 +1,210 @@
+  public readonly websocketServer: websocket.server;
+
+  private hostStore = new HostStore();
+
+  constructor() {
+    this.app.disable('x-powered-by');
+
+    this.httpServer = http.createServer(this.app);
+    this.websocketServer = new websocket.server({
+      httpServer: this.httpServer,
+    });
+
+    this.app
+      .disable('x-powered-by')
+      .get('/', this.onAppGetRoot)
+      .get('/1/user/-/developer-relay/hosts.json', this.lookupHosts.bind(this))
+      .post(
+        '/1/user/-/developer-relay/hosts/:id',
+        this.createConnectionURL.bind(this),
+      );
+
+    this.websocketServer.on('request', this.onRequest.bind(this));
+  }
+
+  public listen(port = 0): number {
+    this.httpServer.listen(port);
+    return (this.httpServer.address() as net.AddressInfo).port;
+  }
+
+  public close(): void {
+    this.httpServer.close();
+  }
+
+  public get port(): number {
+    return (this.httpServer.address() as net.AddressInfo).port;
+  }
+
+  private reject(
+    request: websocket.request,
+    status: number,
+    reason?: string,
+  ): RequestResponse {
+    request.reject(status, reason);
+    return { status };
+  }
+
+  private rejectWithError(
+    request: websocket.request,
+    status: number,
+    message: string,
+  ) {
+    console.error(message);
+    return this.reject(request, status, message);
+  }
+
+  private accept(request: websocket.request, status = 200): RequestResponse {
+    request.accept();
+    return { status };
+  }
+
+  private async onRequest(request: websocket.request) {
+    const headers = request.httpRequest.headers;
+
+    const hostId = headers['x-relay-host-id'] as string;
+    const displayName = headers['x-relay-host-display-name'] as string;
+    const rolesHeader = headers['x-relay-host-roles'] as string;
+    const isHostConnectionAttempted = hostId && displayName;
+
+    const requestedResource =
+      (headers['x-relay-resource'] as string) || request.resource;
+    const peerHostId = requestedResource.split('/')?.[1];
+    const isClientConnectionAttempted = Boolean(peerHostId);
+
+    if (isHostConnectionAttempted && isClientConnectionAttempted) {
+      return this.rejectWithError(
+        request,
+        400,
+        'Ambiguous connection, both authenticated and connection token provided',
+      );
+    }
+
+    if (!isHostConnectionAttempted && !isClientConnectionAttempted) {
+      return this.rejectWithError(
+        request,
+        401,
+        'Unrecognized connection type, neither authenticated nor connection token provided',
+      );
+    }
+
+    if (isHostConnectionAttempted) {
+      let roles: string[];
+
+      try {
+        roles = this.getHostRoles(rolesHeader);
+      } catch (error) {
+        return this.reject(request, 400, (error as Error).message);
+      }
+
+      const host: Host = this.hostStore.addOrReplace({
+        id: hostId,
+        // tslint:disable-next-line:object-shorthand-properties-first
```

Comment:
Why not just do what the linter is asking rather than disable it here?

Reply:
I strongly prefer putting "primary keys" at the top of "queries"

Reply:
The linter is in place to keep everyone's code consistent, so please follow it even if your preferences differ.

Reply:
Oh I'm sorry, I always forget that TSLint rules are __explicitly set by someone__ and am used to overriding them when needed :)  Changed